test/testlib/tls.go

```diff
@@ -0,0 +1,161 @@
+	}
+	for _, expectedFile := range expectedFiles {
+		if !set[expectedFile] {
+			t.Fatalf("Unable to find certificate file %s in path %s", expectedFile, directory)
+		}
+	}
+
+}
+
+func GenerateCustomCertificates(t *testing.T, podName string, namespaceName string, commands []string) {
+	prependCommands := []string{
+		"[ -d " + CERTIFICATES_GENERATION_PATH + " ] && rm -rf " + CERTIFICATES_BACKUP_PATH + " && mv " + CERTIFICATES_GENERATION_PATH + " " + CERTIFICATES_BACKUP_PATH,
+		"rm -rf " + CERTIFICATES_GENERATION_PATH,
+		"mkdir -p " + CERTIFICATES_GENERATION_PATH,
+		"cd " + CERTIFICATES_GENERATION_PATH,
+	}
+	finalCommands := append(prependCommands, commands...)
+	// Execute certificate generation commands
+	executeCommandsInPod(t, podName, namespaceName, finalCommands)
+}
+
+func CopyCertificatesToControlHost(t *testing.T, podName string, namespaceName string) string {
+	options := k8s.NewKubectlOptions("", "")
+	options.Namespace = namespaceName
+
+	prefix := "tls-keys"
+	targetDirectory, err := ioutil.TempDir("", prefix)
+	assert.NilError(t, err, "Unable to create TMP directory with prefix ", prefix)
+	AddTeardown(TEARDOWN_SECRETS, func() { os.RemoveAll(targetDirectory) })
+
+	realTargetDirectory, err := filepath.EvalSymlinks(targetDirectory)
+	assert.NilError(t, err)
+
+	k8s.RunKubectl(t, options, "cp", podName+":"+CERTIFICATES_GENERATION_PATH, realTargetDirectory)
+	verifyCertificateFiles(t, realTargetDirectory)
+
+	return realTargetDirectory
+}
+
+func GenerateTLSConfiguration(t *testing.T, namespaceName string, commands []string, image string) (string, string) {
+	podName := createTLSGeneratorPod(t, namespaceName, image, 30*time.Second)
+	GenerateCustomCertificates(t, podName, namespaceName, commands)
+	keysLocation := CopyCertificatesToControlHost(t, podName, namespaceName)
+
+	CreateSecret(t, namespaceName, CA_CERT_FILE, CA_CERT_SECRET, keysLocation)
+	CreateSecret(t, namespaceName, NUOCMD_FILE, NUOCMD_SECRET, keysLocation)
+	CreateSecretWithPassword(t, namespaceName, KEYSTORE_FILE, KEYSTORE_SECRET, SECRET_PASSWORD, keysLocation)
+	CreateSecretWithPassword(t, namespaceName, TRUSTSTORE_FILE, TRUSTSTORE_SECRET, SECRET_PASSWORD, keysLocation)
+
+	return podName, keysLocation
+}
+
+func RotateTLSCertificates(t *testing.T, options *helm.Options,
+	upgradedOptions *helm.Options, kubectlOptions *k8s.KubectlOptions,
+	initialTLSCommands []string, newTLSCommands []string) (string, string) {
+
+	namespaceName := kubectlOptions.Namespace
+	adminReplicaCount, err := strconv.Atoi(options.SetValues["admin.replicas"])
+	assert.NilError(t, err, "Unable to find/convert admin.replicas value")
+
+	// create initial certs...
+	certGeneratorPodName, _ := GenerateTLSConfiguration(t, namespaceName, initialTLSCommands, "")
+
+	adminReleaseName, namespaceName := StartAdmin(t, options, adminReplicaCount, namespaceName)
+	admin0 := fmt.Sprintf("%s-nuodb-0", adminReleaseName)
+	databaseReleaseName := StartDatabase(t, namespaceName, admin0, options)
```

Comment:
it was originally in the test, however, I thought what if we want to create 3-4 different scenarios with different certificates. All of that would be just boilerplate code. This is why I prefer to have it in a helper function.
I agree that this is not related to the action rotation, so I'm all ears for a new function name instead of `RotateTLSCertificates`.

Reply:
I am not a fan of designing code for the future. It might never happen the way you originally intended. Once we have the actual use case, we can design for that.

Isolating boilerplate in a function is a good idea. It should either live in the test, or it should be a function in the test file (if you think we will reuse it).

This function should only do the rotation (and the AwaitDatabaseUp).

Reply:
Hopefully done.